util/retrieve-cluster-logs.py

```diff
@@ -0,0 +1,213 @@
+import gzip
+import json
+import logging
+import os
+import sys
+import tarfile
+import tempfile
+from datetime import datetime
+
+import argparse
+import boto3
+
+logging.basicConfig(level=logging.INFO)
+LOGGER = logging.getLogger(__file__)
+
+
+def err_and_exit(message):
+    """Log the given error message and exit nonzero."""
+    LOGGER.error(message)
+    sys.exit(1)
+
+
+def get_default_bucket_prefix(cluster_name):
+    """Get default keypath to store logs under in an s3 bucket for the given cluster."""
+    return "{cluster_name}-logs-{end_time}".format(
+        cluster_name=cluster_name, end_time=datetime.now().strftime("%Y%m%d%H%M%S")
+    )
+
+
+def get_log_group(args):
+    """Get log group for args.cluster."""
+    log_group_name = "/aws/parallelcluster/{}".format(args.cluster)
+    logs = boto3.client("logs", region_name=args.region)
+    paginator = logs.get_paginator("describe_log_groups")
+    for result in paginator.paginate(logGroupNamePrefix=log_group_name):
+        for group in result.get("logGroups"):
+            if group.get("logGroupName") == log_group_name:
+                return group
+    err_and_exit(
+        "Unable to find log group for cluster {cluster}. Expected to find one named {log_group_name}".format(
+            cluster=args.cluster, log_group_name=log_group_name
+        )
+    )
+
+
+def verify_bucket_exists_in_region(args):
+    """Verify that bucket exists."""
+    bucket = boto3.resource("s3", region_name=args.region).Bucket(args.bucket)
+    if bucket.creation_date is None:
+        err_and_exit("s3 bucket {} does not exist.".format(args.bucket))
+    bucket_region = bucket.meta.client.get_bucket_location(Bucket=bucket.name).get("LocationConstraint")
```

Comment:
just double check that this works in us-east-1 as well

Reply:
Also a more general comment: we should collect all IAM policies required to run this script and see what happens when the boto3 calls are failing because of permissions. Are we handling this failures?

Reply:
The security/access issues that I handled are:
* the constraint that the bucket and log group must exist in the same region
* the bucket must allow the CloudWatch logs service to get its ACL and put an object where it needs to based on the key prefix used.

There are quite a few boto3 calls that I do not ensure a user has permission for first:
* In addition to the use of the s3 bucket resource to list and download objects, it also call's the s3 client's `get_bucket_location`
* the logs client's `describe_log_groups`, `create_export_task`, `describe_export_tasks` are all called at least once

As of right now, I'm not sure how I would verify a user has permission to make those calls. But I'll look into that today.

Reply:
This did not work in us-east-1. There, the `LocationRestraint` key in the JSON returned by `get_bucket_location` was None instead of the region. Do you know why that is? Is that something I need to worry about for any of the other regions?

Reply:
Never mind my previous comment. I found the information here: https://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region

Reply:
To circle back on my response to the general comment: I looked into what it would take to figure out whether or not a user has the permissions to perform all of the actions listed above.

I think it would be possible to use `iam.simulate_principal_policy` function to do this, but I don't necessarily think the work required to do this well is worth the benfit it gives us over just letting the exceptions bubble up when they occur, given that the error messages tend to be pretty informative for AccessDenied issues.

Reply:
Yes for us-east-1 things are always handled differently there.

For the policy/permissions I think it's enough for now just to be clear in the error message when the script fails due to missing permissions.